core/modules/filter/filter.pages.inc

```diff
@@ -491,28 +491,133 @@ function filter_format_editor_dialog_save($form, &$form_state) {
 
 /**
  * Element validation function.
+ *
+ * Clean up a URL and see if it will pass validation. Instead of stripping
+ * dangerous protocols we flag them as invalid to make the user aware.
+ *
+ * @see filter_format_editor_link_form()
+ * @see filter_format_editor_image_form()
  */
 function _filter_format_editor_link_url_validate(&$element, &$form_state) {
   $value = trim($element['#value']);
 
-  // If Link module is available, automatically clean up URL, automatically
-  // adding http:// or a prefixing slash if needed.
-  if (module_exists('link') && $type = link_validate_url($value)) {
-    $value = link_cleanup_url($value);
-    $form_state['values']['type'] = $type;
-    if ($type === LINK_INTERNAL && strpos($value, '/') !== 0) {
-      $value = '/' . $value;
-    }
+  // Allow empty values, since this will not result in a link tag anyway.
+  if ($value === '') {
+    form_set_value($element, $value, $form_state);
+    return;
   }
 
+  // Clean up the URL so it's more likely to pass validation.
+  $value = _filter_cleanup_url($value);
+
   form_set_value($element, $value, $form_state);
 
-  // Unlike #type = 'url' validation, we allow both relative and absolute paths.
-  if ($value !== '' && !valid_url($value, TRUE) && !valid_url($value, FALSE)) {
+  // Flag if it has any dangerous protocols.
+  $stripped_url = backdrop_strip_dangerous_protocols($value);
+  $dangerous_protocols = $stripped_url !== $value ? TRUE : FALSE;
+
+  // If it has dangerous protocols the URL is invalid. Or if it isn't either
+  // a valid relative nor a valid absolute path.
+  if ($dangerous_protocols || (!valid_url($value, TRUE) && !valid_url($value, FALSE))) {
     form_error($element, t('The URL %url is not valid.', array('%url' => $value)));
   }
 }
 
+/**
+ * Clean up the URL for validation.
+ *
+ * Attempts to create a URL that will pass validation: add a protocol if it
+ * looks like a domain but the protocol is missing; or create a valid, encoded
+ * internal alias.
+ *
+ * @param string $url
+ *
+ * @return string
+ *   A cleaned-up URL that is either a properly formatted external URL or an
+ *   internal path.
+ *
+ * @see _filter_format_editor_link_url_validate()
+ */
+function _filter_cleanup_url($url) {
+  // Will add a default protocol if missing on an external url.
+  $url = _filter_create_valid_external_url($url);
+
+  // If it's an external URL with a protocol we can skip the rest of the
+  // clean up.
+  if (_filter_url_has_protocol($url)) {
+    return $url;
+  }
+
+  // Return anchors as is.
+  if (substr($url, 0, 1) == '#') {
+    return $url;
+  }
+
+  // Return if it begins with a slash to avoid extra language prefixes
+  // if it's a multilingual site.
+  // E.g. /fr/node/1, //domain.tld
```

Comment:
The comment doesn't tell the whole story, though...

It also applies, if a link has been saved before (via autocomplete), because then url has already added the language prefix. - And shouldn't do it again, when editing the link again.

And it's not only about language prefixes, but potentially also about base paths (if Backdrop is installed in a subdirectory).

Reply:
I've edited the comment. I didn't add anything about a base path because I haven't confirmed if that's true.